### ISSUE TITLE
Unable to play h.264 intra-refresh video from the middle of a live stream

### ISSUE THREAD

**Issue by duanyao** [google/ExoPlayer#1654]
## Steps to reproduce
1. Generate a live stream with h.264 intra-refresh in format MPEG-TS over HTTP. E.g. on Windows:
   
   ```
   ffmpeg -f gdigrab -framerate 10 -i desktop -filter:v fps=10 -filter:v scale=640:-1 -f mpegts -muxdelay 0.1 -c:v libx264 -pix_fmt yuv420p -b:v 400K -g 40 -tune zerolatency -x264opts intra-refresh=1:slice-max-size=1300 udp://236.0.0.5:7020?pkt_size=188
   
   "C:\Program Files (x86)\VideoLAN\VLC\vlc.exe" -vvv -I dummy "udp://@236.0.0.5:7020" ":sout=#duplicate{dst=http{mux=ffmpeg{mux=mpegts},dst=:8080/}}"
   ```
2. Play the stream `http://<IP>:8080` with ExoPlayer demo (You need to add an entry); or record part of the stream by `wget` and play the TS file with ExoPlayer demo. I attached a TS file (h264-intra-refresh-middle.ts) and you can play it directly.
## Actual result

The screen of ExoPlayer demo is black, and the playBackState changes from "buffering" to "ended"; the dimension of the video is correct. Aways reproducible.
## Expected result

ExoPlayer demo play the video normally (there may be black screen for several seconds at the beginning, like VLC).
## Other info
- If I disable intra-refresh (remove `-x264opts intra-refresh=1:slice-max-size=1300` from the ffmpeg command line), or record the stream to a file (`a.ts`) from the very beginning, ExoPlayer  demo can play it without problem:

```
ffmpeg -f gdigrab -framerate 10 -i desktop -filter:v fps=10 -filter:v scale=640:-1 -f mpegts -muxdelay 0.1 -c:v libx264 -pix_fmt yuv420p -b:v 400K -g 40 -tune zerolatency -x264opts intra-refresh=1:slice-max-size=1300 pipe: > a.ts
```
## Device

Lenovo 2 A8-50F, Android 5.1.
## ExoPlayer library version

 r1.5.9, from the git master. 
## attachements
- test case: https://github.com/duanyao/ExoPlayer/blob/test/test/h.264-intra-refresh-middle-stream/h264-intra-refresh-middle.ts
- bugreport: https://github.com/duanyao/ExoPlayer/blob/test/test/h.264-intra-refresh-middle-stream/bugreport.txt

**Comment by andrewlewis** [comment]
The stream seems to have one IDR frame at the start, so when playing it starting an some offset (rather than the beginning) we never see an IDR frame, and no frames are output as a result. I don't think we have a simple way to add support for this stream, and it seems like a fairly niche use case.

To make this sort of stream play in ExoPlayer, you can either (1) add IDR frames to the stream (their spacing will determine the worst-case startup latency; note that adding IDR frames defeats one of the purposes of intra refresh, to keep the bitrate and quality reasonably constant) or (2) modify the code in H264Reader to set the synchronization sample flag on the first frame even if it's not an IDR frame. With option (2) you will also need to drop some frames after decoding in the renderer, to avoid showing partially-refreshed frames at the start of playback. The number of frames to drop depends on the intra-refresh interval.

**Comment by duanyao** [comment]
I had tried option (2), but it seems android MediaCodec can't decode such h.264 data (at least not on my devices). Finally I also added ffmpeg's h.264 decoder to ffmpeg-extension and successfully made such stream playable in ExoPlayer. It seems partially-refreshed frames at the start are already dropt by ffmpeg's decoder. Would you like to take a PR?

**Comment by andrewlewis** [comment]
I think MediaCodec will handle this stream fine if you could find a way to queue an IDR frame before the predicted frames. You could try queuing a hard-coded blank IDR frame at the correct resolution before the sample data. You would still need to drop partially-refreshed frames.

What pull request did you have in mind? I think we wouldn't want to make non-trivial changes to support this because it is quite a rare use case. Using FFmpeg for video decoding is going to be resource intensive and battery-inefficient.

**Comment by duanyao** [comment]
The change to H264Reader.java is trivial, just enable `allowNonIdrKeyframes` and set `BUFFER_FLAG_KEY_FRAME` for all samples, see below. 
Most of the work is adding the ffmpeg decoder and a VideoTrackRenderer. Anyway this is an extension, users don't have to include it.

``` java
library/src/main/java/com/google/android/exoplayer/extractor/ts/H264Reader.java 
index be43e71..6f94a42 100644
@@ -172,7 +172,7 @@ import java.util.List;
           output.format(Format.createVideoSampleFormat(null, MimeTypes.VIDEO_H264, null,
               Format.NO_VALUE, Format.NO_VALUE, spsData.width, spsData.height, Format.NO_VALUE,
               initializationData, Format.NO_VALUE, spsData.pixelWidthAspectRatio, null));
-          hasOutputFormat = true;
+          sampleReader.hasOutputFormat = hasOutputFormat = true;
           sampleReader.putSps(spsData);
           sampleReader.putPps(ppsData);
           sps.reset();
@@ -232,6 +232,7 @@ import java.util.List;
     private long samplePosition;
     private long sampleTimeUs;
     private boolean sampleIsKeyframe;
+    private boolean hasOutputFormat;

     public SampleReader(TrackOutput output, boolean allowNonIdrKeyframes,
         boolean detectAccessUnits) {
@@ -422,7 +423,8 @@ import java.util.List;
     }

     private void outputSample(int offset) {
-      int flags = sampleIsKeyframe ? C.BUFFER_FLAG_KEY_FRAME : 0;
+      int flags = (sampleIsKeyframe || allowNonIdrKeyframes && hasOutputFormat) ?
+              C.BUFFER_FLAG_KEY_FRAME : 0;
```

**Comment by hetoan2** [comment]
@duanyao I am interested in your solution to this problem, since I believe the problem I am having has to do with the use of intra-refresh on the stream I am trying to parse. If I could see your full solution and attempt my stream this might help me diagnose my problem further.

**Comment by duanyao** [comment]
@hetoan2 You could try #1682 :)

**Comment by duanyao** [comment]
@andrewlewis I implemented your suggestion with #1682 , it does work for me. However I haven't figured out how to drop partially-refreshed frames yet.
